File: 백준/Gold/1987. 알파벳/알파벳.py
n, m = map(int, input().split())
graph = [list(input().rstrip()) for _ in range(n)]
answer = 0
dy, dx = [0,0,-1,1], [1,-1,0,0]
def dfs(y: int, x: int, path: str):
    global n, m, graph, answer, dy, dx
    answer = max(answer, len(path))
    if not(0 <= y < n and 0 <= x < m):
        return
    for i in range(4):
        ny, nx = y + dy[i], x + dx[i]
        if 0 <= ny < n and 0 <= nx < m and graph[ny][nx] not in path:
            dfs(ny, nx, path + graph[ny][nx])
            
    
dfs(0,0,graph[0][0])
print(answer)

File: 백준/Gold/1987. 알파벳/알파벳.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

class Main {
    static int n, m, answer;
    static int[] dy = { 0, 0, 1, -1 }, dx = { 1, -1, 0, 0 };
    static char[][] board;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] str = br.readLine().split(" ");

        n = Integer.parseInt(str[0]);
        m = Integer.parseInt(str[1]);

        board = new char[n][m];
        for (int i = 0; i < n; i++) {
            String s = br.readLine();
            for (int j = 0; j < m; j++)
                board[i][j] = s.charAt(j);
        }

        dfs(0, 0, (1 << (board[0][0] - 'A')), 1);
        System.out.println(answer);

        br.close();
    }

    static void dfs(int y, int x, int visit, int depth) {
        answer = Math.max(answer, depth);

        for(int i=0 ; i<4 ; i++){
            int ny = y + dy[i], nx = x + dx[i];

            if(0 > ny || ny >= n || 0 > nx || nx >= m )
                continue;

            int idx = board[ny][nx] - 'A';
            if((visit & (1 << idx)) != 0)
                continue;

            dfs(ny, nx, visit | (1 << idx), depth + 1);
        }
    }
}